### PR TITLE
fix: extend deny-misrouted rule to cover /v1/ paths besides batch

### PR DESCRIPTION
**Pull request by bartoszmajsak**
## Description

Follow-up to #881. The `deny-misrouted-model-header` rule excluded all `/v1/` paths from its when-condition, which was intended for the BBR follow-up where `resolvedPath` normalization would handle `/v1/` + header authorization. Without that normalization in place, `/v1/` inference endpoints with a valid model routing header had no authorization rule covering them.

Restores the original batch-path bypass (`/v1/files`, `/v1/batches`) as the deny rule's exclusion instead of the broad `/v1/` exclusion. Batch paths remain authn-only as designed.

Also adds publisher-path test coverage for delegation edge cases (forwarded user without RBAC, forwarded user with only delegate access), authentication layer (no token on publisher path), and response header scoping (batch headers not injected on publisher paths).

Part of [RHOAIENG-69638](https://redhat.atlassian.net/browse/RHOAIENG-69638)

## Merge criteria:

- [x] JIRA(s) are linked in the PR description
- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

[RHOAIENG-69638]: https://redhat.atlassian.net/browse/RHOAIENG-69638?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by CodeRabbit

- **Bug Fixes**
  - Strengthened authorization for publisher model routes to prevent cross-tenant access and block misrouted model headers.
  - Corrected deny vs auth-only handling across chat, files, batches, models, and health endpoints (including proper 403 behavior when applicable).
  - Ensured publisher-path requests without a token return `401` and do not include batch-specific response headers.
  - Fixed namespace-collision and delegated-access cases where required model permissions were missing.
- **Documentation**
  - Updated publisher-path authorization guidance with clarified request-flow behaviors and known limitations.
- **Tests**
  - Expanded end-to-end coverage for the new routing, header, delegation, and namespace collision scenarios.

### CODE REVIEW

File: internal/controller/test/e2e/publisher_path_authpolicy_test.go [hunk 1]
```diff
@@ -211,6 +217,109 @@ func TestEmptyModelHeaderAuthnOnlyFallthrough(t *testing.T) {
...
+		context.Background(),
+		&corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: ns}},
+		metav1.CreateOptions{},
+	)
+	if err != nil {
+		t.Logf("namespace v1 may already exist: %v", err)
+	}
+	t.Cleanup(func() {
+		_ = authEnv.clientset.CoreV1().Namespaces().Delete(context.Background(), ns, metav1.DeleteOptions{})
+	})
```
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply internal/controller/test/e2e/publisher_path_authpolicy_test.go:291]: It's test code, I think we can live with this trade-off considering emphemeral nature of test clusters?

File: internal/controller/test/e2e/publisher_path_authpolicy_test.go [hunk 2]
```diff
@@ -211,6 +217,109 @@ func TestEmptyModelHeaderAuthnOnlyFallthrough(t *testing.T) {
...
+	headers := map[string]string{
+		"x-gateway-model-name": fmt.Sprintf("publishers/%s/models/echo-server", f.ns),
+	}
+	resp, _ := authEnv.gatewayGet(t, "/v1/chat/completions", f.noAccessToken, headers)
+	if resp.StatusCode != http.StatusForbidden {
+		t.Fatalf("expected 403 (/v1/ + model header should be denied without BBR), got %d", resp.StatusCode)
+	}
+}
+
+func TestV1PathWithGarbageModelHeaderDenied(t *testing.T) {
```
Comment (KillianGolds) [inline internal/controller/test/e2e/publisher_path_authpolicy_test.go]: **Tiny naming nit:** the name says `Denied` but the test asserts 200 (documenting the authn-only fallthrough, which the comment explains well). Maybe something like `TestV1PathWithNonPublisherHeaderAuthnOnly` at first glance I thought the test assertion was inverted, again a tiny nit.

File: internal/controller/test/e2e/publisher_path_authpolicy_test.go [hunk 3]
```diff
@@ -211,6 +217,109 @@ func TestEmptyModelHeaderAuthnOnlyFallthrough(t *testing.T) {
...
+	}
+	resp, _ := authEnv.gatewayGet(t, "/v1/chat/completions", f.noAccessToken, headers)
+	if resp.StatusCode != http.StatusForbidden {
+		t.Fatalf("expected 403 (/v1/ + cross-tenant model header should be denied), got %d", resp.StatusCode)
+	}
+}
+
+// --- ns=v1 collision: user with RBAC for garbage SAR identity ---
+
+func TestV1NamespaceCollisionWithRBAC(t *testing.T) {
```
Comment (KillianGolds) [inline internal/controller/test/e2e/publisher_path_authpolicy_test.go:273]: Couple of small robustness thoughts, non-blocking:

- If the create at line 279 fails because ns `v1` already exists, the `t.Cleanup` at line 287 would still delete it right? Might be worth guarding the cleanup on this test having actually created the namespace.
- Back-to-back runs could flake: namespace deletion is async, so a rerun might hit `v1` in Terminating, the create fails, and `createServiceAccount` at line 292 then `Fatalf`s. Waiting for the namespace to be active (or tolerating Terminating with a retry) might help.

Also wondering if the first request at line 307 could assert the expected 200 rather than only logging it, just to pin down the documented authn-only behavior.

**Review by KillianGolds — COMMENTED** [review]
The fix looks right to me. With the broad `/v1/` exclusion from #881, `/v1/` inference endpoints with a valid model routing header seemed to fall through to authn-only, so restoring the narrow batch bypass makes the deny rule fail-closed again. I checked the pre-existing boundary tests (no-header `/v1/`, empty header, participant path + header) and they all seem consistent with the new predicates.

A few small things I noticed, happy to be corrected on any of them:

1. **gofmt**: `internal/controller/test/e2e/publisher_path_authpolicy_test.go` fails `gofmt -l` for me (import ordering at lines 12-14, `corev1` should sort before `rbacv1`). I don't think lint CI catches it since golangci-lint doesn't run with the `e2e` build tag, but probably worth a quick fix to keep the tree gofmt-clean.

2. **Publisher path + mismatched header**: if I'm reading the deny rule's first predicate right (`authpolicy_llm_isvc_userdefined.yaml` lines 107-108), publisher paths are excluded entirely, so `/publishers/A/models/x/...` carrying header `publishers/B/models/y` would be authorized against the *path* (A/x) while routing could act on the *header*. Is routing on publisher paths fully path-pinned after the URLRewrite? If so this is a non-issue, just wanted to check whether the deny rule should also require header == path-derived identity on publisher paths until BBR normalization lands.

3. **Query strings**: as far as I can tell, Envoy's `request.path` includes the query string, so `GET /v1/files?purpose=batch` plus a model header would miss the batch exclusion (`authpolicy_llm_isvc_userdefined.yaml` lines 109-111) and get denied. That is over-deny and fail-closed, so probably fine in practice. But the same pattern on the pre-existing `x-maas-user`/`x-maas-groups` response header predicates (lines 181 and 187) might mean batch identity headers are not injected on `GET /v1/files?purpose=...`. Would `request.url_path` be the better attribute here? Pre-existing either way, so fine as a follow-up if it's a real issue.

File: internal/controller/test/e2e/publisher_path_authpolicy_test.go (hunk 2 above)
Comment (bartoszmajsak) [reply internal/controller/test/e2e/publisher_path_authpolicy_test.go]: Good catch - the name is misleading. Will rename to TestV1PathWithNonPublisherHeaderAuthnOnly.

File: internal/controller/test/e2e/publisher_path_authpolicy_test.go (hunk 3 above)
Comment (bartoszmajsak) [reply internal/controller/test/e2e/publisher_path_authpolicy_test.go:273]: I think coderabbit shared the same concern: https://github.com/opendatahub-io/odh-model-controller/pull/884#discussion_r3602667509

WRT back-to-back flakiness - valid concern. These currently run on ephemeral kind clusters (fresh per CI run), so the namespace collision doesn't happen in practice. If we move toward running these against shared/long-lived clusters, the retry logic would be worth adding. @coderabbitai capture this as a follow-up

> Assert 200 on first request instead of just logging 

Good point. The test logs the result of `/v1/chat/completions` without header but doesn't assert it. Should assert 200 to pin the documented behavior.

Comment (KillianGolds) [reply internal/controller/test/e2e/publisher_path_authpolicy_test.go:273]: Agh yes of course, makes sense, ephemeral kind clusters do sidestep it in practice, and capturing the retry logic as a follow-up for the shared-cluster case works for me. With the 200 assertion added I'm happy here.

File: internal/controller/test/e2e/publisher_path_authpolicy_test.go (hunk 2 above)
Comment (KillianGolds) [reply internal/controller/test/e2e/publisher_path_authpolicy_test.go]: Thanks! Happy to resolve once the rename lands.

**Comment by bartoszmajsak** [comment]
Thanks @KillianGolds, a bit more context:

> 1. **gofmt**: `internal/controller/test/e2e/publisher_path_authpolicy_test.go` fails `gofmt -l` for me (import ordering at lines 12-14, `corev1` should sort before `rbacv1`). I don't think lint CI catches it since golangci-lint doesn't run with the `e2e` build tag, but probably worth a quick fix to keep the tree gofmt-clean.

We should fix linter then, @coderabbitai open issue for this to follow-up.

> 2. **Publisher path + mismatched header**: if I'm reading the deny rule's first predicate right (`authpolicy_llm_isvc_userdefined.yaml` lines 107-108), publisher paths are excluded entirely, so `/publishers/A/models/x/...` carrying header `publishers/B/models/y` would be authorized against the _path_ (A/x) while routing could act on the _header_. Is routing on publisher paths fully path-pinned after the URLRewrite? If so this is a non-issue, just wanted to check whether the deny rule should also require header == path-derived identity on publisher paths until BBR normalization lands.

Good question. Publisher path HTTPRoutes use PathPrefix matching, which should take precedence over the header-only v1-catch-all-model-routing per Gateway API spec. Publisher path routing is path-pinned and the model-access-path rule authorizes against `request.path`, not the header, so a mismatched header doesn't affect the authorization decision. The BBR follow-up's normalization (descoped from this PR) aims to formalize this - it only fires for /v1/ paths, publisher paths pass through unchanged. I hope this makes sense.

> 3. **Query strings**: as far as I can tell, Envoy's `request.path` includes the query string, so `GET /v1/files?purpose=batch` plus a model header would miss the batch exclusion (`authpolicy_llm_isvc_userdefined.yaml` lines 109-111) and get denied. That is over-deny and fail-closed, so probably fine in practice. But the same pattern on the pre-existing `x-maas-user`/`x-maas-groups` response header predicates (lines 181 and 187) might mean batch identity headers are not injected on `GET /v1/files?purpose=...`. Would `request.url_path` be the better attribute here? Pre-existing either way, so fine as a follow-up if it's a real issue.

Sharp eye, it seems to be pre-existing (original template uses the same `request.path` predicates). I just checked and OpenAI files API does use query params (i.e. `?purpose=batch`), so it seems real. Happy to follow up with a switch to `request.url_path` - @coderabbitai do your job.

**Review by KillianGolds — APPROVED** [review]
Thanks for the quick turnaround - rename, import fix, the 200 assertion, and the gist link all addressed, and #887/#888 track the follow-ups. The path-pinning explanation on publisher routes resolved my main concern.

**Comment by bartoszmajsak** [comment]
/override ci/prow/e2e-kserve-llmisvc

autoscaling failures seem unrelated
